errdefs/grpc.go

```diff
@@ -0,0 +1,88 @@
+package errdefs
+
+import (
+	"strings"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// ToGRPC will attempt to map the backend containerd error into a grpc error,
+// using the original error message as a description.
+//
+// Further information may be extracted from certain errors depending on their
+// type.
+//
+// If the error is unmapped, the original error will be returned to be handled
+// by the regular grpc error handling stack.
+func ToGRPC(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	if isGRPCError(err) {
+		// error has already been mapped to grpc
+		return err
+	}
+
+	switch {
+	case IsInvalidArgument(err):
+		return grpc.Errorf(codes.InvalidArgument, err.Error())
+	case IsNotFound(err):
+		return grpc.Errorf(codes.NotFound, err.Error())
+	case IsAlreadyExists(err):
+		return grpc.Errorf(codes.AlreadyExists, err.Error())
+	case IsFailedPrecondition(err):
+		return grpc.Errorf(codes.FailedPrecondition, err.Error())
+	case IsUnavailable(err):
+		return grpc.Errorf(codes.Unavailable, err.Error())
+	}
+
+	return err
+}
+
+// ToGRPCf maps the error to grpc error codes, assembling the formatting string
+// and combining it with the target error string.
+//
+// This is equivalent to errors.ToGRPC(errors.Wrapf(err, format, args...))
+func ToGRPCf(err error, format string, args ...interface{}) error {
+	return ToGRPC(errors.Wrapf(err, format, args...))
+}
+
+func FromGRPC(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	switch grpc.Code(err) {
+	case codes.InvalidArgument:
+		return errors.Wrapf(ErrInvalidArgument, rebaseMessage(ErrInvalidArgument, err))
+	case codes.AlreadyExists:
+		return errors.Wrapf(ErrAlreadyExists, rebaseMessage(ErrAlreadyExists, err))
+	case codes.NotFound:
+		return errors.Wrapf(ErrNotFound, rebaseMessage(ErrNotFound, err))
+	case codes.Unavailable:
+		return errors.Wrapf(ErrUnavailable, rebaseMessage(ErrUnavailable, err))
+	case codes.FailedPrecondition:
+		return errors.Wrapf(ErrFailedPrecondition, rebaseMessage(ErrFailedPrecondition, err))
+	}
+
+	return err
+}
+
+// rebaseMessage removes the repeats for an error at the end of an error
+// string. This will happen when taking an error over grpc then remapping it.
+//
+// Effectively, we just remove the string of cls from the end of err if it
+// appears there.
+func rebaseMessage(cls error, err error) string {
+	// TODO(stevvooe): Maybe lop off the grpcism at the start.
```

Comment:
I added a few testcases and fixed up the handling here. The cause should be preserved in the message now and not repeated if they are unwrapped. A new error, `ErrUnknown` is used for unmappable errors.